### PR TITLE
chore: propagate upstream errors to clients

### LINKED ISSUE #211: chore: gateway propagate status code properly
When querying content, it's always returning 500 status code for all errors.
It should propagate status codes from `indexer`, `cache node` properly.

### PR DESCRIPTION
**Pull request by kckeiks**
Closes #211

### CODE REVIEW

**Review by b0xtch — APPROVED** [review]

File: crates/ursa-gateway/src/worker/cache.rs [hunk 1]
```diff
@@ -136,7 +147,11 @@ impl ServerCache for Cache {
                 ) => body,
                 resp => {
                     error!("Error requested provider: {resp:?}");
-                    bail!("Error requested provider");
+                    let (parts, _) = resp;
```
Comment (scott-dn) [inline crates/ursa-gateway/src/worker/cache.rs:150]:
```suggestion
                (parts, body) => {
                    error!("Error requested provider with parts :{parts:?} and body: {body:?}");
```

File: crates/ursa-gateway/src/util/mod.rs [hunk 2]
```diff
@@ -1 +1,15 @@
+use axum::http::StatusCode;
+
 pub mod timer;
+
+#[derive(Debug)]
+pub enum Error {
```
Comment (scott-dn) [inline crates/ursa-gateway/src/util/mod.rs:6]: maybe move to `util/error.rs`

File: crates/ursa-gateway/src/resolver/mod.rs [hunk 3]
```diff
@@ -129,7 +124,9 @@ impl Resolver {
             .collect();
 
         if provider_addresses.is_empty() {
-            bail!("Failed to get a valid address for provider");
+            return Err(Error::Internal(format!(
```
Comment (scott-dn) [inline crates/ursa-gateway/src/resolver/mod.rs:127]:
```bash
error: useless use of `format!`
   --> crates/ursa-gateway/src/resolver/mod.rs:126:40
    |
126 |               return Err(Error::Internal(format!(
    |  ________________________________________^
127 | |                 "Failed to get a valid address for provider"
128 | |             )));
    | |_____________^ help: consider using `.to_string()`: `"Failed to get a valid address for provider".to_string()`
```

File: crates/ursa-gateway/src/resolver/mod.rs [hunk 4]
```diff
@@ -149,6 +146,6 @@ impl Resolver {
             };
         }
 
-        bail!("Failed to get data")
+        Err(Error::Internal(format!("Failed to get data")))
```
Comment (scott-dn) [inline crates/ursa-gateway/src/resolver/mod.rs:149]:
```bash
error: useless use of `format!`
consider using `.to_string()`: `"Failed to get data".to_string()`
````

File: crates/ursa-gateway/src/worker/cache.rs [hunk 5]
```diff
@@ -126,7 +133,11 @@ impl ServerCache for Cache {
                     error!("Failed to dispatch Fetch command: {e:?}");
                     anyhow!("Failed to dispatch Fetch command")
                 })?;
-            let mut body = match rx.await??.into_parts() {
+            let mut body = match rx
+                .await
+                .map_err(|e| Error::Internal(e.to_string()))??
```
Comment (scott-dn) [inline crates/ursa-gateway/src/worker/cache.rs:138]:
```suggestion
                .map_err(|e| {
                    error!("Failed to receive receive response from resolver: {e:?}");
                    anyhow!("Failed to receive receive response from resolver")
                })??
```

File: crates/ursa-gateway/src/resolver/mod.rs (hunk 3 above)
Comment (kckeiks) [reply crates/ursa-gateway/src/resolver/mod.rs]: Sorry this was copypasta. Updated.

**Review by theBeardA — DISMISSED** [review]
Mint 🍃

File: crates/ursa-gateway/src/resolver/mod.rs [hunk 6]
```diff
@@ -10,6 +10,7 @@ use serde_json::from_slice;
 use tracing::{debug, error};
 
 use crate::resolver::model::ProviderResult;
+use crate::util::error::Error;
```
Comment (scott-dn) [inline crates/ursa-gateway/src/resolver/mod.rs:13]: Merge import

File: crates/ursa-gateway/src/resolver/mod.rs [hunk 7]
```diff
@@ -54,19 +55,12 @@ impl Resolver {
...
-                },
-                ..,
-            ) => {
-                error!("Error requested indexer - Got 404: {endpoint}");
-                bail!("Error requested indexer - Got 404: {endpoint}")
-            }
-            resp => {
-                error!("Error requested indexer: {endpoint} {resp:?}");
-                bail!("Error requested indexer: {endpoint}")
+            (parts, _) => {
```
Comment (scott-dn) [inline crates/ursa-gateway/src/resolver/mod.rs:58]: We may want to log the body too.

File: crates/ursa-gateway/src/server/route/api/v1/get.rs [hunk 8]
```diff
@@ -10,6 +10,7 @@ use cid::Cid;
 use serde_json::{json, Value};
 use tokio::sync::RwLock;
 
+use crate::util::error::Error;
```
Comment (scott-dn) [inline crates/ursa-gateway/src/server/route/api/v1/get.rs:13]: Merge import

File: crates/ursa-gateway/src/worker/cache.rs [hunk 9]
```diff
@@ -17,6 +17,7 @@ use tokio_util::io::ReaderStream;
 use tracing::{error, info, log::warn};
 
 use crate::cache::{ByteSize, Tlrfu};
+use crate::util::error::Error;
```
Comment (scott-dn) [inline crates/ursa-gateway/src/worker/cache.rs:20]: Merge import

File: crates/ursa-gateway/src/worker/cache.rs [hunk 10]
```diff
@@ -126,17 +133,27 @@ impl ServerCache for Cache {
                     error!("Failed to dispatch Fetch command: {e:?}");
                     anyhow!("Failed to dispatch Fetch command")
                 })?;
-            let mut body = match rx.await??.into_parts() {
+            let mut body = match rx
+                .await
+                .map_err(|e| {
+                    error!("Failed to receive receive response from resolver: {e:?}");
```
Comment (scott-dn) [inline crates/ursa-gateway/src/worker/cache.rs:139]: Typo?

**Review by b0xtch — APPROVED** [review]

**Pull request merged**